docs/src/proposals/partitioned-inflationary-rewards-distribution.md

```diff
@@ -0,0 +1,103 @@
+current epoch, but must also be random across different epochs. One way to
+achieve these properties is to hash the account's pubkey with some epoch
+dependent values, sort the results, and divide them into `M` bins. The epoch
+dependent value can be the epoch number, total rewards for the epoch, the leader
+pubkey for the epoch block, etc. `M` can be choses based on 50K account per
+block, which equal to `ceil(num_stake_accounts/50,000)`.
+
+In order to avoid putting extra burden of computing and credit the stake reward
+for blocks produced during the `rewarding interval`, we can reduce the compute
+budget limits on those blocks in `rewarding interval`, and reserve some computing
+and read/write capacity to perform stake rewarding.
+
+### Challenges
+
+1. stake accounts reads/writes during the `rewarding interval`
+
+`epoch_start..epoch_start+N+M` Because of the delayed credit of the rewards,
+Reads to those stake accounts may not will return the value that the user are
+expecting (viz. not include the recent epoch stake rewards). Writes to those
+stake accounts will be lost once the reward are credited on block
+`epoch_start+N+M`. We will need to modify the runtime to restrict read/writes to
+stake accounts during the `rewarding interval`. Any transactions, which involves
+stake accounts, will result in a new execution error, i.e. "stake rewards
+pending, account access is restricted". However, normal rpc queries, such as
+'getBalance', will return the current lamport of the account. The user can
+expect the rewards to be credit as some time point during the 'rewarding
+interval'.
+
+2. snapshot taken during the `rewarding interval`
+
+If a snapshot is taken during the `rewarding interval`, it would miss the
+rewards for the stake accounts. Any plain restart from those snapshots will be
+wrong, unless we reconstruct the rewards from the recent epoch boundary. This
+will add some complexity to validator restart. In the first implementation, we
+will force *not* taking any snapshot and *not* performing accounts hash
+calculation during the `rewarding interval`. In future, if needed, we can
+revisit to enable taking snapshots and perform hash calculation during reward
+interval.
+
+3. account-db related action during the `rewarding interval`
+
+Account-db related action such as flush, clean, squash, shrink etc. may touch
+and evict the stake accounts from account db's cache during the `rewarding
+interval`. This will slow down the credit in the future at bank `epoch_start+N`.
+We may need to exclude such account_db actions for stake_accounts during
+`rewarding interval`. This is going to be a performance tuning problem. In the
+first implementation, for simplicity, we will keep the account-db action as it
+is, and make the `credit interval` larger to accommodate the performance hit
+when writing back those accounts. In future, we can continue tuning account db
+actions during 'rewarding interval'.
```

Comment:
Hopefully this shouldn't be a big perf hit since clean/flush/shrink are only working on rooted slots, and these rewards computations are near the tip. Only contention would be brief lock on the accounts index for those pubkeys being credited